src/main/java/ch/njol/skript/effects/EffVisualEffect.java

```diff
@@ -77,50 +80,57 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
 		effects = (Expression<VisualEffect>) exprs[base];
 		direction = (Expression<Direction>) exprs[base + 1];
 		where = exprs[base + 2];
-		if (exprs[base + 3] != null) {
-			if (exprs[3].getReturnType() == Player.class)
-				players = (Expression<Player>) exprs[base + 3];
-			else
-				radius = (Expression<Number>) exprs[base + 3];
-		}
+		players = (Expression<Player>) exprs[base + 3];
+		radius = (Expression<Number>) exprs[base + 4];
+
 		if (effects instanceof Literal) {
-			final VisualEffect[] effs = effects.getAll(null);
-			boolean hasLocationEffect = false, hasEntityEffect = false;
-			for (final VisualEffect e : effs) {
-				if (e.isEntityEffect())
+			//noinspection ConstantConditions
+			VisualEffect[] effs = effects.getAll(null);
+
+			boolean hasLocationEffect = false;
+			boolean hasEntityEffect = false;
+			for (VisualEffect e : effs) {
+				if (e.getType().isEntityEffect())
 					hasEntityEffect = true;
 				else
 					hasLocationEffect = true;
 			}
+
 			if (!hasLocationEffect && players != null)
 				Skript.warning("Entity effects are visible to all players");
```

Comment:
Unrelated to the PR but isn't this warning misleading, like, why is it a warning to show effects to all players?

Reply:
Because this warning would show up when specifying which players to show the effect to, e.g. `play wolf hearts on the clicked wolf to player`: since `wolf hearts` is an entity effect, it'll be visible to all players no matter what, so specifying which players this should be visible to is useless.